vamb/__main__.py

```diff
@@ -939,6 +980,306 @@ def run(
+            mmseq_line = mmseq_map.get(r['contigs'], '').split(';')
+        except AttributeError:
+            preds.append(';'.join(pred_line))
+            continue
+        if mmseq_line[0] != '':
+            for i in range(len(mmseq_line)):
+                if i < len(pred_line):
+                    pred_line[i] = mmseq_line[i]
+        preds.append(';'.join(pred_line))
+    df_gt[f'predictions'] = preds
+
+    PREDICTED_PATH = vamb_options.out_dir.joinpath("results_taxonomy_predictor.csv")
+    df_gt.to_csv(PREDICTED_PATH, index=None)
+
+    log(f"\nCompleted taxonomy predictions in {round(time.time() - begintime, 2)} seconds.", logfile, 0)
+
+
+def run_vaevae(
+    vamb_options: VambOptions,
+    comp_options: CompositionOptions,
+    abundance_options: AbundanceOptions,
+    vae_training_options: VAETrainingOptions,
+    predictor_training_options: VAETrainingOptions,
+    taxonomy_options: TaxonomyOptions,
+    cluster_options: ClusterOptions,
+    logfile: IO[str],
+):
+    log("Starting Vamb version " + ".".join(map(str, vamb.__version__)), logfile)
+    log("Date and time is " + str(datetime.datetime.now()), logfile, 1)
+    begintime = time.time()
+
+    # Get TNFs, save as npz
+    composition = calc_tnf(comp_options, vamb_options.out_dir, logfile)
+
+    # Parse BAMs, save as npz
+    abundance = calc_rpkm(
+        abundance_options,
+        vamb_options.out_dir,
+        composition.metadata,
+        vamb_options.n_threads,
+        logfile,
+    )
+
+    time_generating_input = round(time.time() - begintime, 2)
+    log(
+        f"\nTNF and coabundances generated in {time_generating_input} seconds.",
+        logfile,
+        1,
+    )
+
+    tnfs, lengths = composition.matrix, composition.metadata.lengths
+    rpkms = abundance.matrix
+
+    if taxonomy_options.taxonomy_predictions_path:
+        df_gt = pd.read_csv(taxonomy_options.taxonomy_predictions_path)
+        graph_column = df_gt[f'predictions']
+        indices_mmseq = range(len(df_gt))
+    elif taxonomy_options.taxonomy_path and not taxonomy_options.no_predictor:
+        run_taxonomy_predictor(
+            vamb_options=vamb_options,
+            comp_options=comp_options,
+            abundance_options=abundance_options,
+            taxonomy_options=taxonomy_options,
+            predictor_training_options=predictor_training_options,
+            logfile=logfile,
+        )
+        PREDICTED_PATH = vamb_options.out_dir.joinpath("results_taxonomy_predictor.csv")
+        df_gt = pd.read_csv(PREDICTED_PATH)
+        graph_column = df_gt[f'predictions']
+        indices_mmseq = range(len(df_gt))
+    elif taxonomy_options.no_predictor:
+        df_mmseq = pd.read_csv(taxonomy_options.taxonomy_path, delimiter='\t', header=None)
+        df_mmseq = df_mmseq[~(df_mmseq[2] == 'no rank')]
+        graph_column = df_mmseq[8]
+        ind_map = {c: i for i, c in enumerate(composition.metadata.identifiers)}
+        indices_mmseq = [ind_map[c] for c in df_mmseq[0]]
```

Comment:
Again, there looks to be some code duplicated with `run_taxonomy_predictor`, which suggests this could potentially be refactored.